include/drivers/dai.h

```diff
@@ -0,0 +1,717 @@
+ * @param channels Number of words per frame.
+ * @param format Data stream format as defined by DAI_FMT_* constants.
+ * @param options Configuration options as defined by DAI_OPT_* constants.
+ * @param frame_clk_freq Frame clock (WS) frequency, this is sampling rate.
+ * @param mem_slab memory slab to store RX/TX data.
+ * @param block_size Size of one RX/TX memory block (buffer) in bytes.
+ * @param timeout Read/Write timeout. Number of milliseconds to wait in case TX
+ *        queue is full or RX queue is empty, or 0, or SYS_FOREVER_MS.
+ */
+struct dai_legacy_params {
+	uint8_t word_size;
+	uint8_t channels;
+	dai_fmt_t format;
+	dai_opt_t options;
+	uint32_t frame_clk_freq;
+	struct k_mem_slab *mem_slab;
+	size_t block_size;
+	int32_t timeout;
+};
+
+/** Main dai config struct */
+struct dai_config {
+	enum dai_type type;
+	uint32_t dai_index;
+
+	/* generic dai config members */
+	dai_fmt_t format;
+	uint32_t channels;
+	uint32_t frame_clk_freq;
+	uint8_t group_id;
+	uint8_t flags;
+
+	/* optional dai specific params */
+	const void *params;
+};
+
+/* TODO 1) array of clocks 2) wallclk naming to be clear */
+struct dai_timestamp_cfg {
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+	int type; /* SSP, DMIC, HDA, etc. */
+	int direction; /* Playback, capture */
+	int index; /* For SSPx to select correct timestamp register */
+	int dma_id; /* DMA instance id */
+	int dma_chan_index; /* Used DMA channel */
+	int dma_chan_count; /* Channels in single DMA */
+};
+
+struct dai_timestamp_data {
+	uint64_t walclk; /* Wall clock */
+	uint64_t sample; /* Sample count */
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * For internal use only, skip these in public documentation.
+ */
+__subsystem struct dai_driver_api {
+	int (*init)(const struct device *dev);
+	int (*deinit)(const struct device *dev);
+
+	int (*config_set)(const struct device *dev, const struct dai_config *cfg,
+			  const void * bespoke_cfg);
+	const struct dai_config *(*config_get)(const struct device *dev,
+					       enum dai_dir dir);
+
+	uint32_t (*get_property)(const struct device *dev, enum dai_dir dir, int stream_id,
+				 enum dai_property prop);
+
+	int (*trigger)(const struct device *dev, enum dai_dir dir, int stream_id,
+		       enum dai_trigger_cmd cmd);
+
+	/* simple mem_block interface */
+	int (*read)(const struct device *dev, void **mem_block, size_t *size, int stream_id);
+	int (*write)(const struct device *dev, void *mem_block, size_t size, int stream_id);
+
+	/* timestamp ops */
+	int (*timestamps_config)(const struct device *dev, struct dai_timestamp_cfg *cfg);
+	int (*timestamp_start)(const struct device *dev, struct dai_timestamp_cfg *cfg);
+	int (*timestamp_stop)(const struct device *dev, struct dai_timestamp_cfg *cfg);
+	int (*timestamp_get)(const struct device *dev, struct dai_timestamp_cfg *cfg,
+			     struct dai_timestamp_data *tsd);
```

Comment:
We need to comment if mandatory or optional

Reply:
i.e. ```/* timestamp ops - optional */```

Reply:
ok